New picoDst branch with FMS hits

plexoos opened the pull request.

pliuphys committed:
Added time to StPicoEvent
Save time when generating picodst from mudst and recover when reading
picodst, so St_db_Maker has timestamp.

pliuphys committed:
Introducing FMS hit class StPicoFmsHit
Note: StPicoArray and StPicoDstMaker code to be modified to include hits in the PicoDst

pliuphys committed:
Enabled FMS hit IO

pliuphys committed:
makePicoDst.C: Updates for FMS

plexoos commented on StPicoEvent/StPicoEvent.h:
```diff
@@ -110,6 +111,8 @@ class StPicoEvent : public TObject
   UShort_t mFillId;
   Float_t  mBField;
 
+  Int_t    mTime;//To set timestamp for St_db_Maker. This is what StMuDstMaker used to GetEvtHddr()->SetGMTime(cast into unsigned int). Peifeng Liu
```
@pliuphys 
Do we ever use event timestamp? If we are talking about calibration and other corrections, these are usually done on per run basis. St_db_Maker knows extracts correct data from the DB based on the run number, we don't need to provide any additional timestamps in this workflow

pliuphys replied:
I found that if I don't set a timestamp in code reading picodst and then use the St_db_Maker, the code would crash with St_db_Maker complaining no timestamp. So should I just set a somehow arbitrary timestamp (like 20160101,0) in the macro?

Also, I don't see any mechanism for the current PicoDst to send runid information to other makers. Should that be implemented?

plexoos replied:
Could you prepare a small example to demonstrate this? I am not surprised that St_db_Maker complains when it does not have any reference to the time but it is surprising that adding the time to StPicoEvent solves this problem.

I also don't see how St_db_Maker would pick up the right time in this macro https://github.com/star-bnl/star-picoDst/blob/master/makePicoDst.C
Perhaps, it is a good question for the picoDst-hn mailing list and those to wrote the macro.

pliuphys replied:
In St_db_Maker::Make(), GetDateTime is invoked,

const TDatime &St_db_Maker::GetDateTime() const
{
  if (!fIsDBTime) return StMaker::GetDateTime();
  //(void )printf("**** fIsDBTime is set, returning its value\n");
  return fDBTime;
}

And then, StMaker::GetDateTime() uses the event header, which is written by MuDst but not (official) PicoDst.

const TDatime  &StMaker::GetDateTime() const 
{    
   const StEvtHddr *hd = GetEvtHddr();
   return hd->GetDateTime();
}

I guess I should create a separate repository and upload test case(s)?

plexoos commented on StPicoEvent/StPicoFmsHit.h:
```diff
@@ -0,0 +1,72 @@
+
+ private:
+  UShort_t mChannelDetectorId;//32*channel+detectorId. Allow channel 0-2047, detectorId 0-31.
+  UShort_t mAdc;
```
I thought we agreed to save both ADC and energy information for the hit

plexoos commented on StPicoDstMaker/StPicoDstMaker.cxx:
```diff
@@ -1175,3 +1183,37 @@ bool StPicoDstMaker::selectVertex()
+    LOG_ERROR << "No StEvent found" << endm;
+    return;
+  }
+  StFmsCollection* fmsColl = event->fmsCollection();
```
Don't we have the same collection available in muDst branches? If yes, get it from muDst instead of StEvent

plexoos commented on StPicoDstMaker/StPicoDstMaker.cxx:
```diff
@@ -1175,3 +1183,37 @@ bool StPicoDstMaker::selectVertex()
+void StPicoDstMaker::fillFmsHits()
+{
+  /*
+  StFmsDbMaker* fmsDb=static_cast<StFmsDbMaker*>(GetMaker("fmsDb"));//can save this into a member variable
```
Yes, this would be a better place to access StFmsDbMaker or St_db_Maker if you need to update the FMS hit data before storing them

plexoos commented on StPicoEvent/StPicoFmsHit.cxx:
```diff
@@ -0,0 +1,61 @@
+    return;
+
+  fmsDbMaker=static_cast<StFmsDbMaker*>(StMaker::GetChain()->GetMaker("fmsDb"));
+
```
It is quite counter intuitive to have StFmsDbMaker integrated in this StPicoFmsHit class. The "hit" should be essentially a data structure. By adding the dependence on StFmsDbMaker in the Hit class you force users to always instantiate it even if they just want to access the FMS hit data without doing any manipulation with it.

If really needed it is better to access the DB at the StPicoDstMaker level instead of here

pliuphys commented:
In St_db_Maker::Make(), GetDateTime is invoked,

const TDatime &St_db_Maker::GetDateTime() const
{
  if (!fIsDBTime) return StMaker::GetDateTime();
  //(void )printf("**** fIsDBTime is set, returning its value\n");
  return fDBTime;
}

And then, StMaker::GetDateTime() uses the event header, which is written by
MuDst but not (official) PicoDst.

const TDatime  &StMaker::GetDateTime() const
{
   const StEvtHddr *hd = GetEvtHddr();
   return hd->GetDateTime();
}

I guess I should create a separate repository and upload test case(s)?

On Thu, Jun 1, 2017 at 2:16 PM, Dmitri Smirnov <notifications@github.com>
wrote:

> *@plexoos* commented on this pull request.
> ------------------------------
>
> In StPicoEvent/StPicoEvent.h
> <https://github.com/star-bnl/star-picoDst/pull/70#discussion_r119690607>:
>
> > @@ -110,6 +111,8 @@ class StPicoEvent : public TObject
>    UShort_t mFillId;
>    Float_t  mBField;
>
> +  Int_t    mTime;//To set timestamp for St_db_Maker. This is what StMuDstMaker used to GetEvtHddr()->SetGMTime(cast into unsigned int). Peifeng Liu
>
> Could you prepare a small example to demonstrate this? I am not surprised
> that St_db_Maker complains when it does not have any reference to the time
> but it is surprising that adding the time to StPicoEvent solves this
> problem.
>
> I also don't see how St_db_Maker would pick up the right time in this
> macro https://github.com/star-bnl/star-picoDst/blob/master/makePicoDst.C
> Perhaps, it is a good question for the picoDst-hn mailing list and those
> to wrote the macro.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/star-bnl/star-picoDst/pull/70#discussion_r119690607>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AYPiwXYj0k_3YAAx4ohpjaEvC1ZRxyphks5r_wAPgaJpZM4NtQ56>
> .
>

plexoos committed:
Removed dependence on StEvent in StPicoDstMaker::fillFmsHits()

plexoos committed:
cmake: Added include dir for FMS

plexoos committed:
makePicoDst.C: Removed dependence on StEvent and StFmsHitMaker

plexoos committed:
Added StPicoFmsFiller to fill FMS hit branch in picoDst

The information about FMS hits comes from the corresponding branch in muDst

plexoos committed:
Use StPicoFmsFiller to fill FMS branch in picoDst

If the fmsDbMaker is available and the FMS hit branch in muDst is empty (i.e.
has zero entries) the hits are recovered from the corresponding data block in
StTriggerData. This is done to recover FMS hits in those muDst files in which
FMS hit branch was not filled as in Run 16 production.

plexoos committed:
StPicoFmsHit: [Cosmetic] Add braces to avoid compiler warning

plexoos committed:
cmake: Added StPicoFmsHit to the library

plexoos committed:
Added dependencies for testing

plexoos committed:
makePicoDst.C: Prefixed loaded libs with lib

plexoos committed:
makePicoDst.C: Enable FMS branches when reading muDst

plexoos committed:
Minor clean-up

plexoos committed:
Corrected code formatting with 'astyle -s2 -p -H -A3 -k3 -O -o -y -Y -f'

plexoos committed:
StPicoFmsHit: Removed dependency on StFmsDbMaker

plexoos committed:
minor comments

plexoos committed:
Consistency with other picoEvent classes

plexoos committed:
StPicoFmsHit: Print to std::cout stream
